### PR TITLE
Running multiple nodes on the same host

### PR DESCRIPTION
**Pull request by bokobza**
This PR contains a few commits allowing multiple nodes to run on the same machine:

* Make sure the node terminates:
If an exception is thrown (for example the DBreeze fails to initialize), the ManualResetEventSlim is never set and the node stays waiting for it. So if for example a second node is started on a machine, it will fail whilst creating the DBreeze db but the process will keep running. This fix allows the process to terminate.

* Checking for null before disposing of Dbreeze

* Allow setting of custom data directory.
Previously, trying to start a node with a directory different than the default one would throw an exception. Now we create a directory after the one specified in the datadir argument.

* Split of ApiUri
Splitting the apiUri between ApiUri and ApiPort so that it's easier to set up a port without having to worry about setting the host.
For example, a user starting the node could just start the node with -apiport=37555.

* Removed the NOASSEMBLYCONTEXT constant from csproj files. 

Helps the following:
 https://github.com/stratisproject/Breeze/issues/322
https://github.com/stratisproject/Breeze/issues/314
https://github.com/stratisproject/StratisBitcoinFullNode/issues/806
https://github.com/stratisproject/StratisBitcoinFullNode/issues/257

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 1]
```diff
@@ -9,9 +9,21 @@ namespace Stratis.Bitcoin.Features.Api
     /// </summary>
     public class ApiSettings
     {
+        /// <summary>The default port used by the API when the node runs on the bitcoin network.</summary>
+        public const int DefaultBitcoinApiPort = 37220;
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: I think this should go in the Network class (next to rpc port)

File: src/Stratis.Bitcoin/Configuration/NodeSettings.cs [hunk 2]
```diff
@@ -174,14 +174,21 @@ public NodeSettings LoadArguments(string[] args)
...
                 this.DataDir = this.CreateDefaultDataDirectories(Path.Combine("StratisNode", this.Name), this.Network);
             }
-
-            if (!Directory.Exists(this.DataDir))
-                throw new ConfigurationException($"Data directory {this.DataDir} does not exist.");
-
+            else
+            {
+                // Create the data directories if they don't exist.
+                string directoryPath = Path.Combine(dataDir, this.Name, this.Network.Name);
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Configuration/NodeSettings.cs]: So now the path folder structure is
```
- data-dir (or default)
    - coin-name
        -network
            - main 
            - test
```

Right?

File: src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs [hunk 3]
```diff
@@ -34,21 +34,29 @@ public static class FullNodeExtensions
...
+
                 Console.CancelKeyPress += (sender, eventArgs) =>
                 {
                     shutdown();
                     // Don't terminate the process immediately, wait for the Main thread to exit gracefully.
                     eventArgs.Cancel = true;
                 };
 
-                await node.RunAsync(cts.Token, "Application started. Press Ctrl+C to shut down.", "Application stopped.").ConfigureAwait(false);
+                try
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: If there is a try catch here can we remove the one in program.cs?

**Comment by dangershony** [comment]
I think this also fixes https://github.com/stratisproject/StratisBitcoinFullNode/issues/804

File: src/Stratis.Bitcoin.Api.Tests/ApiSettingsTest.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            // Assume the default values for the static flags.
+            Transaction.TimeStamp = false;
+            Block.BlockSignature = false;
+        }
+
+        /// <summary>
+        /// Tests that if no API settings are passed and we're on the bitcoin network, the defaults settings are used.
+        /// </summary>
+        [Fact]
+        public void GivenNoApiSettingsAreProvidedAndOnBitcoinNetworkThenDefaultSettingAreUsed()
```
[the remark this answers is not in the document]
Comment (bokobza) [reply src/Stratis.Bitcoin.Api.Tests/ApiSettingsTest.cs]: will change

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 1 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: This is only for the API feature though. It does not really have an obvious place in the Network class.

File: src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs (hunk 3 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: Ah this is to invoke the `done.Set();`
What about putting a finally here and let the exception in program.cs be the only place to log errors?

```
                 try
                 {
                     await node.RunAsync(cts.Token, "Application started. Press Ctrl+C to shut down.", "Application stopped.").ConfigureAwait(false);
                 }
                 finally
                 {
                     done.Set();
                 }
```

**Pull request merged**